### ISSUE TITLE
Yolo Onnx Export -> Experiencing Worse Accuracy

### ISSUE THREAD

**Issue by Budelon** [Deci-AI/super-gradients#1952]
### 🐛 Describe the bug

i have trained my the object detection model Yolo Nas S on my own dataset, that includes only one class.

When i test the results, i have F1=0.9, which is satisfatory for me. 

Code: 
best_model = models.get(Models.YOLO_NAS_S, num_classes=NUM_CLASSES, checkpoint_path=checkpointPath)
prediction = best_model.predict(image, fuse_model=False, conf=CONF, iou=IOU)
bboxes = prediction.prediction.bboxes_xyxy
class_names = prediction.class_names
labels = prediction.prediction.labels
confidence = prediction.prediction.confidence
image = prediction.image.copy()

But, when i run the export model code below:
model = models.get(Models.YOLO_NAS_S, checkpoint_path=model_checkpoint, num_classes=1, )
export_result = model.export("model/240405i8.onnx", 
                            output_predictions_format=DetectionOutputFormatMode.FLAT_FORMAT)

and estimate the F1 with the code below with the ONNX model:
session = onnxruntime.InferenceSession(
        onnx_model_path, 
        providers=["CPUExecutionProvider"],
                                        #    providers=["CPUExecutionProvider", "CUDAExecutionProvider"]
                                        )
    inputs = [o.name for o in session.get_inputs()]
    outputs = [o.name for o in session.get_outputs()]
image_640 = cv2.resize(image, (640, 640))
        image_bchw = np.transpose(np.expand_dims(image_640, 0), (0, 3, 1, 2))
        prediction = session.run(outputs, {inputs[0]: image_bchw})
        precictions = [x for x in prediction[0] if x[5] > CONF]
        bboxes = [[
            int(x[1] * image.shape[1] / 640),
            int(x[2] * image.shape[0] / 640),
            int(x[3] * image.shape[1] / 640),
            int(x[4] * image.shape[0] / 640)
        ] for x in precictions]
        confidence = [x[5] for x in precictions]
        labels = [x[6] for x in precictions]
        class_names = classes

The F1 with the same confidence of 0.8 decay from 0.9 to 0.05. 

Summary: 
NOT ONNX model: F1 = 0.9
ONNX model: F1 = 0.05
*Same validation dataset.

### Versions

Collecting environment information...
PyTorch version: 2.2.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.19 (main, Mar 21 2024, 17:11:28)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-26-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Laptop GPU
Nvidia driver version: 535.161.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             16
On-line CPU(s) list:                0-15
Vendor ID:                          GenuineIntel
Model name:                         11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
CPU family:                         6
Model:                              141
Thread(s) per core:                 2
Core(s) per socket:                 8
Socket(s):                          1
Stepping:                           1
CPU max MHz:                        4600,0000
CPU min MHz:                        800,0000
BogoMIPS:                           4608.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear ibt flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          384 KiB (8 instances)
L1i cache:                          256 KiB (8 instances)
L2 cache:                           10 MiB (8 instances)
L3 cache:                           24 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-15
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] onnx==1.15.0
[pip3] onnx-graphsurgeon==0.3.27
[pip3] onnxruntime==1.15.0
[pip3] onnxsim==0.4.36
[pip3] torch==2.2.2
[pip3] torchmetrics==0.8.0
[pip3] torchvision==0.17.2
[pip3] triton==2.2.0
[conda] numpy                     1.23.0                   pypi_0    pypi
[conda] torch                     2.2.2                    pypi_0    pypi
[conda] torchmetrics              0.8.0                    pypi_0    pypi
[conda] torchvision               0.17.2                   pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi

**Comment by BloodAxe** [comment]
In general this should not happen. Can you try visualizing  predictions for a single image for non-exported and exported model? Please attach the complete inference code you are using for onnx model.

**Comment by BloodAxe** [comment]
The code that you provided looks normal to my eye. Unfortunately it is not enough to tell what is going on. Is it possible that you provide a minimal reproducing example that demonstrate the issue end-to-end?What it should contain:
1) A trained model checkpoint file (.pth)
2) A single image that you are seeing discrepancy in the results
3) An inference code  that uses `model.predict()` and `model.export() + onnxruntime inference` to demonstate the difference in predictions.

**Comment by Budelon** [comment]
> Such a big drop can suggest a few things:
> 
> 1. the input images are not as expected. i.e. the model was trained on pixel values [0,255], but the ONNX model is fed [0,1].
> 2. the post-processing or metric is wrong - you said that you can visualize the images, do the boxes look ok? Do you get many boxes or only a few? And do the boxes you get look correctly positioned? is the classification ok?

Answer: 
2) The model confidence i use is the same for both models -> Before ONNX export and after export. So, when i run both models in the same image with the same confidence, the bounding box is detected correctly with "non exported model" and is not detected at all in the "exported model"... When i decrease the confidence to smaller values, then it starts to detect bounding boxes.

**Comment by Budelon** [comment]
> The code that you provided looks normal to my eye. Unfortunately it is not enough to tell what is going on. Is it possible that you provide a minimal reproducing example that demonstrate the issue end-to-end?What it should contain:
> 
> 1. A trained model checkpoint file (.pth)
> 2. A single image that you are seeing discrepancy in the results
> 3. An inference code  that uses `model.predict()` and `model.export() + onnxruntime inference` to demonstate the difference in predictions.

Sure, you can get in this[ link](https://drive.google.com/drive/folders/1H0EWDF-eNPFguzWa4qSGVanyg_84PNNC?usp=sharing).

**Comment by BloodAxe** [comment]
Ok, just by looking at the example image I can immediately see the issue.

Your preprocessing  method resize input image to 640x64 without preserving an aspect ratio of the input image. The example image that you have shared seems to be a single crop around a person and when resized to 640x640 human in picture is resized to the extreme and this confuses a model.

So what you actually want to do when doing ONNX inferencing is to mimic the validation preprocessing.
Meaning you should preserve aspect ratio of the images to avoid appearing of non-realistic objects.
The way to go is this:
1) Resize image to have longest max size of 640 pixels
2) Pad image to 640x640

Then do inference as usual, and as a last step - undo preprocessing by shifting (if center padding was used) and scaling boxes back to original image coordinates.

**Comment by Budelon** [comment]
> So what you actually want to do when doing ONNX inferencing is to mimic the validation preprocessing.

Thanks a lot! It worked, i added a np.zeros.. in order to pad the image... 

I'm using supergradients DetectionDataset class to create my own dataset to train. How can i check all operations that must be done to mimic the validation preprocessing?

**Comment by BloodAxe** [comment]
It depends. But if you are using default dataset params, you can check it here

https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/recipes/dataset_params/coco_detection_yolo_nas_dataset_params.yaml#L143

**Comment by Budelon** [comment]
> It depends. But if you are using default dataset params, you can check it here
> 
> https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/recipes/dataset_params/coco_detection_yolo_nas_dataset_params.yaml#L143

Thanks!!
